sound/soc/soc-pcm.c

```diff
@@ -2165,38 +2184,50 @@ int dpcm_be_dai_trigger(struct snd_soc_pcm_runtime *fe, int stream,
 			    (be->dpcm[stream].state != SND_SOC_DPCM_STATE_PAUSED))
 				goto next;
 
-			if (!snd_soc_dpcm_can_be_free_stop(fe, be, stream))
+			if (be->dpcm[stream].state == SND_SOC_DPCM_STATE_START)
+				be->dpcm[stream].be_start--;
+
+			if (be->dpcm[stream].be_start != 0)
 				goto next;
 
 			ret = soc_pcm_trigger(be_substream, cmd);
-			if (ret)
+			if (ret) {
+				if (be->dpcm[stream].state == SND_SOC_DPCM_STATE_START)
+					be->dpcm[stream].be_start++;
```

Comment:
I'm wondering is this safe as a replacement for snd_soc_dpcm_can_be_free_stop(). In old code, BE trigger is not called in this case at all, but in new code trigger is called but failure is expected if the BE was busy. OTOH, in old code BE received many triggers in any case, so maybe this is ok as well.

Reply:
Had to @plbossart re-read my comment a few times as well. :) So in existing/old code, if snd_soc_dpcm_can_be_free_stop() returns false (i.e. no connected FEs are in running/paused/suspended state), snd_pcm_trigger() would NOT be called at all in this case. Instead we "continue" and jump to next BE.

With new/updated code in this PR, I understand it is possible we would call soc_pcm_trigger() in this case, but it would return an error and we'd take the branch on L2192 (and move to next one). 

Are all BEs prepared to handle such trigger call? Or did I interpret this wrong, and in all cases where snd_soc_dpcm_can_be_free_stop() would return false, in new code be_start count goes to zero and we jump to next BE with new code as well.

Reply:
Yes it's the latter interpretation @kv2019i. Rather than checking the current FE states, the assumption was that these FE states would have triggered a prior transition bringing the refcount to zero. I really don't see how we could have a BE in 'START' state if all FEs are not generating any data.

Reply:
Hmm, ok, this seems correct. Fine for me.